src/meta-srv/src/region/supervisor.rs

```diff
@@ -0,0 +1,580 @@
+pub struct RegionSupervisor {
+    /// Used to detect the failure of regions.
+    failure_detector: RegionFailureDetector,
+    /// The interval of tick
+    tick_interval: Duration,
+    /// Receives [Event]s.
+    receiver: Receiver<Event>,
+    /// [Event] Sender.
+    sender: Sender<Event>,
+    /// The context of [`SelectorRef`]
+    selector_context: SelectorContext,
+    /// Candidate node selector.
+    selector: SelectorRef,
+    /// Region migration manager.
+    region_migration_manager: RegionMigrationManagerRef,
+    // TODO(weny): find a better way
+    kv_backend: KvBackendRef,
+    table_metadata_manager: TableMetadataManagerRef,
+}
+
+/// [`HeartbeatAcceptor`] forwards heartbeats to [`RegionSupervisor`].
+pub(crate) struct HeartbeatAcceptor {
+    sender: Sender<Event>,
+}
+
+impl HeartbeatAcceptor {
+    /// Accepts heartbeats from datanodes.
+    pub(crate) async fn accept(&self, heartbeat: DatanodeHeartbeat) {
+        if let Err(e) = self.sender.send(Event::HeartbeatArrived(heartbeat)).await {
+            error!(e; "RegionSupervisor is stop receiving heartbeat");
+        }
+    }
+}
+
+#[cfg(test)]
+impl RegionSupervisor {
+    /// Returns the [Event] sender.
+    pub(crate) fn sender(&self) -> Sender<Event> {
+        self.sender.clone()
+    }
+}
+
+impl RegionSupervisor {
+    pub(crate) fn new(
+        options: PhiAccrualFailureDetectorOptions,
+        tick_interval: Duration,
+        selector_context: SelectorContext,
+        selector: SelectorRef,
+        region_migration_manager: RegionMigrationManagerRef,
+        kv_backend: KvBackendRef,
+    ) -> Self {
+        let (tx, rx) = tokio::sync::mpsc::channel(1024);
+        Self {
+            failure_detector: RegionFailureDetector::new(options),
+            tick_interval,
+            receiver: rx,
+            sender: tx,
+            selector_context,
+            selector,
+            region_migration_manager,
+            table_metadata_manager: Arc::new(TableMetadataManager::new(kv_backend.clone())),
+            kv_backend,
+        }
+    }
+
+    /// Returns the [`HeartbeatAcceptor`].
+    pub(crate) fn heartbeat_acceptor(&self) -> HeartbeatAcceptor {
+        HeartbeatAcceptor {
+            sender: self.sender.clone(),
+        }
+    }
+
+    /// Returns the [`RegionSupervisorTicker`].
+    pub(crate) fn ticker(&self) -> RegionSupervisorTickerRef {
+        Arc::new(RegionSupervisorTicker {
+            tick_interval: self.tick_interval,
+            sender: self.sender.clone(),
+            tick_handle: Mutex::new(None),
+        })
+    }
+
+    /// Runs the main loop.
+    pub(crate) async fn run(&mut self) {
+        while let Some(event) = self.receiver.recv().await {
+            match event {
+                Event::Tick => {
+                    let regions = self.detect_region_failure();
+                    self.handle_region_failures(regions).await;
```

Comment:
Will executing handle_region_failures block the processing of other events?

Reply:
Nope, the procedure will be spawned to background runtime